dvc/commands/data.py

```diff
@@ -0,0 +1,364 @@
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+) -> Dict[str, List[str]]:
+    # TODO: add support for shallow diff in dvc-data
+    # TODO: we may want to recursively do in_cache check
+    d = {}
+
+    from dvc_data.objects.tree import Tree
+
+    if isinstance(new_obj, Tree):
+        root = os.path.sep.join([root, ""])
+
+    if not _in_cache(old_obj, cache):
+        d["not_in_cache"] = [root]
+
+    if old_obj is None and new_obj is None:
+        return d
+    if old_obj is None:
+        return {"added": [root], **d}
+    if new_obj is None:
+        return {"deleted": [root], **d}
+    if old_obj.hash_info != new_obj.hash_info:
+        return {"modified": [root], **d}
+    return {"unchanged": [root], **d}
+
+
+def granular_diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+) -> Dict[str, List[str]]:
+    from dvc_data.diff import ROOT
+    from dvc_data.diff import diff as odiff
+    from dvc_data.objects.tree import Tree
+
+    def path_join(root: str, *paths: str) -> str:
+        if paths == ROOT:
+            return root
+        return os.path.sep.join([root, *paths])
+
+    diff_data = odiff(old_obj, new_obj, cache)
+    drop_root = isinstance(new_obj, Tree)
+
+    output: Dict[str, List[str]] = defaultdict(list)
+    for state in ("added", "deleted", "modified", "unchanged"):
+        items = getattr(diff_data, state)
+        output[state].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not (drop_root and item.new.key == ROOT)
+        )
+        # TODO: PERF: diff is checking not_in_cache for each even if we only
+        # need it for the index.
+        # BUG: not_in_cache file also shows up as modified in staged and
+        # unstaged. We currently don't know if it is really modified.
+        output["not_in_cache"].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not item.old.in_cache
+            and not (drop_root and item.new.key == ROOT)
+            and state != "added"
+        )
+    return output
+
+
+def diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+    granular: bool = False,
+) -> Dict[str, List[str]]:
+    if granular:
+        return granular_diff(root, old_obj, new_obj, cache)
+    return shallow_diff(root, old_obj, new_obj, cache)
+
+
+class CmdDataStatus(CmdBase):
+    COLORS = {
+        "not_in_cache": "red",
+        "unstaged": "yellow",
+        "staged": "green",
+        "untracked": "cyan",
+    }
+    LABELS = {
+        "not_in_cache": "Not in cache",
+        "staged": "DVC committed changes",
+        "unstaged": "DVC uncommitted changes",
+        "untracked": "Untracked files",
+        "unmodified": "DVC unmodified files",
+    }
+    JSON_LABELS = {
+        "staged": "dvc-committed",
```

Comment:
I slightly prefer `dvc-committed`/`untracked` to `committed`/`untracked` because the first one at least hints that `untracked` is not `dvc-untracked`. However, it's not a strong opinion, especially for JSON since users of it probably already have a good idea of what each label means. I'm fine with whatever you decide.

Reply:
`dvc` is redundant, will remove from JSON output. I am open for reconsideration.

Reply:
As for stdout, it's better to specify "DVC-(un)tracked)" since there's also Git-(un)tracked. Clearly DVC status commands shouldn't report about what Git is doing but there's a lot of similar functionality/terminology here, and users can get confused.

Reply:
Technically, `DVC-untracked == Git-untracked`. For others it makes sense, and we have `DVC` suffix added to them.